### ISSUE TITLE
[Bug]: Metal prefill GEMM is not token-exact vs CPU on near-tie logits (S >= GEMM_MIN); decode unaffected

### ISSUE THREAD

**Issue by lBroth** [JustVugg/colibri#622]
On tiny fixture containers, teacher-forced prefill disagrees CPU vs Metal at 2 of 32 positions, deterministically, on stock `dev` at 7b03ce0 (M5 Pro, macOS 26.5.2). Greedy decode is token-identical on the same containers. This came out of the #587 hardware runs; filing it separately as agreed there, since it reproduces on stock dev where neither PR's code exists.

Reproduce, from `c/` with a tiny oracle snapshot in place (`tools/make_glm_oracle.py`, needs the transformers>=5.11 / Python 3.12 venv; the converter additionally needs torch and safetensors). One trap to know: if `--indir` points at a missing or empty snapshot the converter prints `converted 0 main shard(s)` and exits 0, and every command below then passes vacuously on an empty container, so check that line says 1.

```
python3 tools/convert_fp8_to_int4.py --indir glm_tiny --outdir /tmp/t_i4g64 --ebits 4 --io-bits 8 --group-size 64 --n-layers 5
python3 tools/convert_fp8_to_int4.py --indir glm_tiny --outdir /tmp/t_i3 --ebits 4 --io-bits 8 --xbits 3 --group-size 0 --n-layers 5

SNAP=/tmp/t_i4g64 TF=1 ./colibri 64 4 16                 # CPU reference
SNAP=/tmp/t_i4g64 COLI_METAL=1 TF=1 ./colibri 64 4 16    # Metal
# same CPU/Metal pair with SNAP=/tmp/t_i3 for the second container
```

Observed, identical across repeated runs:

| container | positions that differ (CPU token vs Metal token) |
|---|---|
| grouped int4 g64 | pos 14: 34 vs 29, pos 27: 7 vs 112 |
| int3 experts, per-row int4 dense | pos 19: 91 vs 103, pos 22 vs 23 shift |

The culprit is localized, not guessed. `COLI_METAL_GEMM_MIN=1000` (keep every `matmul_qt` GEMM on the CPU, default threshold is 16 rows) makes both containers bit-identical to the CPU reference, both directions of the discriminator. It also explains why greedy stays token-identical here: this harness prefills a 12-token prompt, below the 16-row threshold, and decode steps are S=1, so the dense matmuls never leave the CPU on that path. A 16-plus-token prompt would route its prefill through the same GEMM, so greedy is not structurally immune, it just never crosses the threshold in this harness. The MoE and fused-attention paths are exonerated by the same discriminator rather than by absence: on the g64 container the batched-expert counters read zero, on the int3 container the shared expert does run on the GPU (2 blocks) yet the output is bit-identical once the GEMM stays on the CPU, and METAL-ATTN never engages on toy dims.

To be clear about what this is and is not. The GEMM kernel is numerically fine, `make metal-test` passes it at ~3e-06 against a 1e-04 tolerance. But `docs/metal.md` says "Token-exact vs the CPU path", and on prefill that only holds while no two logits sit closer than the GEMM's accumulation-order drift. These 4-bit toy containers are pathological, near-ties everywhere, so they surface it at 2 in 32 positions; a real container should cross it far more rarely, but the promise as written has a hole, and teacher-forced comparisons against an oracle (the workflow the repo itself documents) inherit it. Options as I see them, in increasing cost: scope the docs claim to the decode path, or make the prefill GEMM's accumulation order stable against the CPU loop; the fused decode paths already behave token-exactly on these same fixtures, so exactness there is evidently achievable. The top-2 logit-gap instrumentation being stood up on #587 will put a number on the margins at exactly these positions, so I would wait for that before choosing.

One small thing found while measuring: the oracle harness prints `run: TF=1 DEBUG_LOGITS=1 for top-5 logit dump`, but no such flag exists in the source, the string is the only occurrence.

Happy to re-run any of this on demand, the fixture side takes seconds here.

**Comment by JustVugg** [comment]
This is an exemplary bug report — the diagnosis is right and I verified every piece of it against `dev`:

- `g_metal_gemm_min = 16` (default), overridable by `COLI_METAL_GEMM_MIN` — so a ≥16-token prefill routes `matmul_qt` through the GPU GEMM, and S=1 decode plus a 12-token prompt never do. That matches exactly why decode stays token-exact here and prefill doesn't.
- `COLI_METAL_GEMM_MIN=1000` making both containers bit-identical to CPU localizes it to the GEMM cleanly, and exonerates MoE/fused-attention.
- `DEBUG_LOGITS` really is a dead string — one occurrence, the oracle mismatch message, no flag reads it.

**What this is:** not a kernel defect (the GEMM passes `make metal-test` at ~3e-6 vs a 1e-4 tolerance) but a GPU-vs-CPU **accumulation-order** difference that only becomes visible when two logits sit inside that drift — i.e. near-ties. The 4-bit toy containers are pathological (near-ties everywhere), so they surface it at 2/32; a real container should cross it far more rarely. Decode is unaffected. So the concrete defect is that `docs/metal.md` promises "token-exact vs the CPU path" without the prefill/near-tie caveat, and the oracle workflow the repo documents inherits that hole.

**Plan, matching your cost ordering:**

1. **Now — scope the docs claim + fix the dead flag.** I'll correct `docs/metal.md` to say decode is token-exact and prefill's GPU GEMM can diverge on near-tie logits by accumulation order, with `COLI_METAL_GEMM_MIN=<large>` as the bit-exact escape hatch (and the fmt=4 oracle pattern for teacher-forced comparisons). And I'll fix the `DEBUG_LOGITS` line — since it's genuinely useful for exactly this kind of near-tie debugging, I'd rather **implement** a real top-5 logit dump than delete the hint; if that balloons, I'll just remove the string.

2. **Later — stable-order prefill GEMM.** The fused decode paths already reduce in a CPU-stable order on these same fixtures, so exactness is achievable; it's a real fix but costs kernel work / possibly some GEMM throughput. I agree with waiting for the top-2 logit-gap instrumentation from #587 to put a number on how often a *real* container actually crosses the margin before spending that. If it's genuinely rare, the docs-scoping is the right level of fix.

**One connection worth flagging:** this is exactly the confound in verifying **#457 / #587** (Metal fmt=4) token-exactness on tiny fixtures — a fmt=4 mismatch there could be this GEMM near-tie drift rather than the new kernel. Those PRs' correctness checks should run with `COLI_METAL_GEMM_MIN=1000` (or on a real container) to isolate the fmt=4 kernel from this pre-existing prefill drift. I'll note that on #587.

Thanks for filing it separately against stock `dev` — that's what makes it actionable. I'll put up the docs + `DEBUG_LOGITS` fix shortly.

**Comment by JustVugg** [comment]
Landed on `dev` (#624). Two things you flagged are fixed:

- **`docs/metal.md`** now scopes the claim: decode is token-exact; prefill's GPU GEMM can diverge on near-tie logits by accumulation order (not a kernel bug), with `COLI_METAL_GEMM_MIN=100000` documented as the bit-exact escape hatch for teacher-forced comparisons.
- **`DEBUG_LOGITS` is now real** (it was a dead string). On a teacher-forcing mismatch it prints the top-5 logits, the **top1-top2 margin**, and the expected/got tokens' logits, e.g.:
  ```
  [LOGITS] pos=5 top1-top2 gap=2.449e-03 | expected=35 (-0.23799)  got=34 (1.73265) | top5: 34:1.73265 197:1.73020 ...
  ```
  Run `SNAP=<container> COLI_METAL=1 TF=1 DEBUG_LOGITS=1 ./colibri 64 4 16` on your Mac and it'll print exactly how close the ties are at pos 14/27 — I'd expect sub-3e-6 gaps, which is the GEMM drift.

The stable-accumulation-order kernel fix is deferred as you suggested, pending the #587 top-2 logit-gap instrumentation to quantify how often a real container crosses the margin. Thanks again — the report was textbook. Leaving this open until you confirm the docs read right and the dump shows what you expect on the Mac.

**Comment by lBroth** [comment]
Confirmed on current dev at c031307, both halves.

The docs read right. The scoped claim, the escape hatch and the DEBUG_LOGITS pointer in `docs/metal.md` match what the fixture shows, and the flag works as described.

The dump at the two flip positions corrects one expectation though. You wrote you'd expect sub-3e-6 gaps there; measured on the gs64 fixture the flips sit far above that: pos 14 has a CPU-side top1-top2 gap of 3.26e-3 (and it is really a triple tie, the top three logits sit within 1.3e-2 of each other), pos 27 sits at 5.77e-3. So the effective drift at the logits is on the order of 1e-3, not 3e-6. That is not a contradiction of the diagnosis, it is the difference between the single-GEMM relative error and what reaches the logits after the perturbed prefill state propagates through the remaining layers. It matches what the #587 instrumentation is finding at real scale too, where the cross-backend drift is macroscopic everywhere and flips land wherever a trajectory's gap dips under it.

Practical consequence for the deferred kernel fix: the crossing-rate question should be priced against a drift band of roughly 1e-3 and up at the logits, not the kernel's 3e-6. Fine to close on my side, and thanks for turning both flags around the same night.

**Comment by JustVugg** [comment]
Closing: confirmed resolved on current dev at c031307, both halves, by the reporter.

The scoped claim, the escape hatch and the `DEBUG_LOGITS` pointer in `docs/metal.md` now match what the fixture actually shows.

Worth recording why this closed as documentation rather than as a kernel change. Metal's prefill GEMM diverging from CPU on near-tie logits is a real property of a different accumulation order, not a defect to be patched away, and decode was never affected. The right fix was to say so precisely, scope where token-exactness is promised, and give people a way to check. Thanks for verifying both halves instead of just the code half.
